src/lib/libast/man/strexpr.3
.TH STREXPR 3
.SH NAME
strexpr \- long integer arithmetic expression evaluator
.SH SYNOPSIS
.nf
.ft CR
long strexpr(const char* \fIs\fP, char** \fIe\fP,
             long (*\fIconv\fP)(const char* cs, char** ce, void *handle),
             void *\fIhandle\fP)
.ft
.fi
.SH DESCRIPTION
.I strexpr
evaluates the long integer arithmetic expression in the null-terminated string
.I s
and returns the result.
If
.I e
is not 0 then
.I *e
is set to point to the first unknown character in the expression.
.PP
If
.I conv
is not 0, then it is called when an unknown token is encountered in
.IR s .
.I cs
points to the beginning of the unknown token.
The return value is the long integer value of the unknown token and
.I ce
must be set to point to the first character after the unknown token.
The
.I handle
pointer argument to
.I strexpr
is passed to
.I conv
when called; its purpose is user-defined.
If an expression syntax error is encountered, then
.I conv
is called with
.I cs
set to 0 and
.I *ce
pointing to the error message text.
.PP
In addition to the normal C expressions and integer constant styles,
numbers in any base
.I b
<= 2 <= 64
may be represented as
.IR b # nnnn .
For bases greater than 10, the notation is as described in
.IR strton (3).
.SH "SEE ALSO"
.IR strton (3)


NEWS
This documents significant changes in the dev branch of ksh 93u+m.
For full details, see the git log at: https://github.com/ksh93/ksh
Uppercase BUG_* IDs are shell bug IDs as used by the Modernish shell library.

2026-09-18:

- Fixed: a command substitution will now correctly capture the output of
  background processes launched within it but not explicitly waited for.

2026-09-17:

- Fixed multiple memory leaks occurring upon subshell exit when assigning a
  value to a variable with a shell discipline function within a subshell.

- Fixed a crash occurring when assigning, within a subshell, an invalid
  value to a variable of a type defined by the 'enum' command.

2026-09-16:

- Fixed: when the special IFS (field splitting characters) variable was
  assigned to within a ${ shared-state command substitution }, the state of IFS
  failed to persist beyond it, and a use after free (potential crash) occurred.

- Fixed bounds checking when assigning to fixed-size arrays (declared like,
  e.g., 'typeset -a array[4]'). Assigning to an out-of-range subscript now
  produces the expected error instead of potentially crashing.

2026-09-15:

- Fixed a memory leak when assigning, within a subshell, a value to a
  special variable (one of IFS, FPATH, CDPATH, SHELL, ENV, VISUAL, EDITOR,
  HISTFILE, OPTIND, SECONDS, RANDOM,.sh.match, .sh.version, LINENO, _,
  LC_CTYPE, LC_NUMERIC, LC_COLLATE, LC_MESSAGES, LC_ALL, and LANG).

2026-09-13:

- Fixed a memory leak that occurred when converting an existing scalar
  variable to an associative array using 'typeset -A'.

2026-09-12:

- Implemented advisory locking for $HISTFILE (~/.sh_history by default) so
  that multiple ksh instances can write to the history file without risking
  data loss when one of them decides to trim the file. Note that ksh
  versions before 2026-09-12 will ignore the lock, so the risk continues to
  exist if an older ksh binary shares write access to the file.

- History file trimming no longer requires write access to the directory in
  which the history file resides and will no longer change the directory's
  last-modified timestamp. However, it now requires a usable /tmp or $TMPDIR.

- I/O errors while writing to the history file are no longer silently ignored.

- Fixed a possible crash when attempting to set an integer variable to a
  multibyte character.

2026-09-10:

- Fixed a rare crash that could occur if a signal (such as interrupt, Ctrl+C)
  occurred while cleaning up the state of a virtual subshell upon its exit.

2026-09-08:

- Fixed: PWD and ENV were not properly reinitialised when executing a shell
  script without a #! path.

- Fixed a possible crash after unsetting RANDOM. Bug introduced on 2022-07-21.

2026-09-04:

- Fixed a longstanding bug in which output data loss or corruption could
  intermittently occur if, within a command substitution, standard error is
  redirected to standard output and, within that redirection, standard
  output is then connected to a pipe.

2026-09-02:

- Fixed a bug, introduced on 2025-05-30, that made 'print -s' ineffective
  at writing to the history file. This functionality is now restored.

2026-08-31:

- Fixed a crash in the expr(1) path-bound built-in command that could occur
  when using the non-standard 'match' or 'index' keyword.

2026-08-30:

- Fixed a bug that caused non-importable environment variables (i.e., those
  with names that are not valid shell variable names) to disappear from the
  environment of child processes after invoking more than two external
  commands, potentially crashing the shell. Bug introduced on 2025-05-30.

2026-08-29:

- If an error (such as a syntax error) occurs in a trap action, the error
  message will now have the word "(trap)" in the call stack for clarity.

2026-08-26:

- Following up to 2026-08-16, fixed a remaining problem in the 'read'
  built-in with combined -d and -n options when reading from a pipe; the
  custom delimiter is now correctly used instead of newline.

2026-08-16:

- Fixed a long-standing bug in the 'read' built-in: the -d option was
  ignored when combined with -n. The two options can now be used together.

2026-08-14:

- In the --man/--help output for built-in commands, text marked as italic
  is no longer rendered as bold + underlined, but only as underlined.

2026-08-13:

- Fixed corruption of $_, crashing the shell upon 'typeset -p _', after an
  error occurs trying to redefine a type defined with 'typeset -T'.

2026-08-06:

- Fixed ksh crashing upon parsing a very long list of simple commands
  separated by semicolons or, in 'eval' or dot script contexts, by newlines.

- [v1.1] The -n/--noexec option can now effectively be set in subshells of
  interactive shells, e.g., within a parentheses block.

- [v1.1] The 'eval' command has a new -n/--noexec option that causes it to
  subject the command(s) given to static analysis in the manner of the
  -n/--noexec shell option, emitting warnings for deprecated usages.

2026-08-04:

- Fixed a potential crash when suspending (^Z), on the interactive shell,
  a multi-element pipeline with the last element running in the current
  environment.

- For non-interactive shells, the COLUMNS and LINES variables now initially
  respect their values inherited from the environment, if any. They will
  still be updated upon the next window size change (SIGWINCH) or, on the
  interactive shell, when any external command returns. This change aligns
  ksh with bash in supporting idioms like:
	$ COLUMNS=60 ksh somescript.ksh
	$ COLUMNS=60 print --man

2026-08-03:

- Fixed crashes that could occur when an arithmetic syntax error is thrown
  in an arithmetic expression processed via the print or printf built-ins.

- ksh no longer crashes or misbehaves when an arithmetic expression
  longer than 32767 bytes contains an array subscript.

2026-08-01:

- Fixed bogus expansion values of ${var/#pattern/replacement} or
  ${var/%pattern/replacement} that occurred when var is unset.


src/lib/libast/string/strexpr.c
/***********************************************************************
*                                                                      *
*               This software is part of the ast package               *
*          Copyright (c) 1985-2011 AT&T Intellectual Property          *
*          Copyright (c) 2020-2026 Contributors to ksh 93u+m           *
*                      and is licensed under the                       *
*                 Eclipse Public License, Version 2.0                  *
*                                                                      *
*                A copy of the License is available at                 *
*      https://www.eclipse.org/org/documents/epl-2.0/EPL-2.0.html      *
*         (with md5 checksum 84283fa8859daf213bdda5a9f8d1be1d)         *
*                                                                      *
*                 Glenn Fowler <gsf@research.att.com>                  *
*                  David Korn <dgk@research.att.com>                   *
*                   Phong Vo <kpv@research.att.com>                    *
*                  Martijn Dekker <martijn@inlv.org>                   *
*            Johnothan King <johnothanking@protonmail.com>             *
*                                                                      *
***********************************************************************/
/*
 * G. S. Fowler
 * D. G. Korn
 * AT&T Bell Laboratories
 *
 * long integer arithmetic expression evaluator
 * long constants may be represented as:
 *
 *	0ooo		octal
 *	0[xX]hhh	hexadecimal
 *	ddd		decimal
 *	n#ccc		base n, 2 <= b <= 64
 *
 * NOTE: all operands are evaluated as both the parse
 *	 and evaluation are done on the fly
 */

#include <ast.h>
#include <ctype.h>

#define getchr(ex)	(*(ex)->nextchr++)
#define peekchr(ex)	(*(ex)->nextchr)
#define ungetchr(ex)	((ex)->nextchr--)

#define err(ex,msg)	return seterror(ex,msg)

typedef struct				/* expression handle		*/
{
	char*		nextchr;	/* next expression char		*/
	char*		errchr;		/* next char after error	*/
	char*		errmsg;		/* error message text		*/
	long		(*convert)(const char*, char**, void*);
	void*		handle;		/* user convert handle		*/
} Expr_t;

/*
 * set error message string
 */

static COLD long
seterror(Expr_t* ex, char* msg)
{
	if (!ex->errmsg) ex->errmsg = msg;
	ex->errchr = ex->nextchr;
	ex->nextchr = "";
	return 0;
}

/*
 * evaluate a subexpression with precedence
 */

static long
expr(Expr_t* ex, int precedence)
{
	int	c;
	long	n;
	long	x;
	char*	pos;
	int	operand = 1;

	while (c = getchr(ex), isspace(c));
	switch (c)
	{
	case 0:
		ungetchr(ex);
		if (!precedence) return 0;
		err(ex, "more tokens expected");
	case '-':
		n = -expr(ex, 13);
		break;
	case '+':
		n = expr(ex, 13);
		break;
	case '!':
		n = !expr(ex, 13);
		break;
	case '~':
		n = ~expr(ex, 13);
		break;
	default:
		ungetchr(ex);
		n = 0;
		operand = 0;
		break;
	}
	for (;;)
	{
		switch (c = getchr(ex))
		{
		case 0:
			goto done;
		case ')':
			if (!precedence) err(ex, "too many )'s");
			goto done;
		case '(':
			n = expr(ex, 1);
			if (getchr(ex) != ')')
			{
				ungetchr(ex);
				err(ex, "closing ) expected");
			}
		gotoperand:
			if (operand) err(ex, "operator expected");
			operand = 1;
			continue;
		case '?':
			if (precedence > 1) goto done;
			if (peekchr(ex) == ':')
			{
				getchr(ex);
				x = expr(ex, 2);
				if (!n) n = x;
			}
			else
			{
				x = expr(ex, 2);
				if (getchr(ex) != ':')
				{
					ungetchr(ex);
					err(ex, ": expected for ? operator");
				}
				if (n)
				{
					n = x;
					expr(ex, 2);
				}
				else n = expr(ex, 2);
			}
			break;
		case ':':
			goto done;
		case '|':
			if (peekchr(ex) == '|')
			{
				if (precedence > 2) goto done;
				getchr(ex);
				x = expr(ex, 3);
				n = n || x;
			}
			else
			{
				if (precedence > 4) goto done;
				x = expr(ex, 5);
				n |= x;
			}
			break;
		case '^':
			if (precedence > 5) goto done;
			x = expr(ex, 6);
			n ^= x;
			break;
		case '&':
			if (peekchr(ex) == '&')
			{
				if (precedence > 3) goto done;
				getchr(ex);
				x = expr(ex, 4);
				n = n && x;
			}
			else
			{
				if (precedence > 6) goto done;
				x = expr(ex, 7);
				n &= x;
			}
			break;
		case '=':
		case '!':
			if (peekchr(ex) != '=') err(ex, "operator syntax error");
			if (precedence > 7) goto done;
			getchr(ex);
			x = expr(ex, 8);
			if (c == '=') n = n == x;
			else n = n != x;
			break;
		case '<':
		case '>':
			if (peekchr(ex) == c)
			{
				if (precedence > 9) goto done;
				getchr(ex);
				x = expr(ex, 10);
				if (c == '<') n <<= x;
				else n >>= x;
			}
			else
			{
				if (precedence > 8) goto done;
				if (peekchr(ex) == '=')
				{
					getchr(ex);
					x = expr(ex, 9);
					if (c == '<') n = n <= x;
					else n = n >= x;
				}
				else
				{
					x = expr(ex, 9);
					if (c == '<') n = n < x;
					else n = n > x;
				}
			}
			break;
		case '+':
		case '-':
			if (precedence > 10) goto done;
			x = expr(ex, 11);
			if (c == '+') n +=  x;
			else n -= x;
			break;
		case '*':
		case '/':
		case '%':
			if (precedence > 11) goto done;
			x = expr(ex, 12);
			if (c == '*') n *= x;
			else if (x == 0) err(ex, "divide by zero");
			else if (c == '/') n /= x;
			else n %= x;
			break;
		default:
			if (isspace(c)) continue;
			pos = --ex->nextchr;
			if (isdigit(c)) n = strton(ex->nextchr, &ex->nextchr, NULL, 0);
			else if (ex->convert) n = (*ex->convert)(ex->nextchr, &ex->nextchr, ex->handle);
			if (ex->nextchr == pos) err(ex, "syntax error");
			goto gotoperand;
		}
		if (ex->errmsg) return 0;
		if (!operand) err(ex, "operand expected");
	}
 done:
	ungetchr(ex);
	if (!operand) err(ex, "operand expected");
	return n;
}

/*
 * evaluate an integer arithmetic expression in s
 *
 * (long)(*convert)(const char* string, char** end, void* handle)
 * is a user supplied conversion routine that is called when unknown
 * chars are encountered; string points to the part to be
 * converted and end is adjusted to point to the next non-converted
 * character; if string is 0 then end points to an error message string
 *
 * NOTE: (*convert)() may call strexpr(ex, )
 */

long
strexpr(const char* s, char** end, long(*convert)(const char*, char**, void*), void* handle)
{
	long	n;
	Expr_t	ex;

	ex.nextchr = (char*)s;
	ex.errmsg = 0;
	ex.convert = convert;
	ex.handle = handle;
	n = expr(&ex, 0);
	if (peekchr(&ex) == ':')
		seterror(&ex, "invalid use of :");
	if (ex.errmsg)
	{
		if (convert) (*convert)(NULL, &ex.errmsg, handle);
		ex.nextchr = ex.errchr;
		n = 0;
	}
	if (end) *end = ex.nextchr;
	return n;
}
